src/tiled/objecttemplatemodel.cpp

```diff
@@ -0,0 +1,109 @@
+/*
+ * objecttemplatemodel.cpp
+ * Copyright 2017, Thorbjørn Lindeijer <thorbjorn@lindeijer.nl>
+ * Copyright 2017, Mohamed Thabet <thabetx@gmail.com>
+ *
+ * This file is part of Tiled.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "objecttemplatemodel.h"
+
+#include "templategroup.h"
+
+namespace Tiled {
+
+ObjectTemplateModel::ObjectTemplateModel(const QList<TemplateGroup*> &templateGroups, QObject *parent):
+    QAbstractItemModel(parent),
+    mTemplateGroups(templateGroups)
+{
+}
+
+QModelIndex ObjectTemplateModel::index(int row, int column,
+                                       const QModelIndex &parent) const
+{
+    if (parent == QModelIndex())
+        return createIndex(row, column, mTemplateGroups.at(row));
+
+    if (TemplateGroup *templateGroup = toTemplateGroup(parent)) {
+        if (row < templateGroup->templateCount())
+            return createIndex(row, column, templateGroup->templateAt(row));
+    }
+
+    return QModelIndex();
+}
+
+QModelIndex ObjectTemplateModel::parent(const QModelIndex &index) const
+{
+    if (ObjectTemplate *objectTemplate = toObjectTemplate(index))
+        return createIndex(0, 0, objectTemplate->templateGroup());
```

Comment:
Row 0 can't be always right, you need to determine the index of the template group in `mTemplateGroups`.